rfd/0008-api.md

```diff
@@ -0,0 +1,241 @@
+    authservice.pb.go
+    ```
+
+#### Move client implementation to `api/client`
+
+Move relevant lib/client.*Client code to `api/client`.
+Track dependecies, make sure that new `api/client` module has no dependencies
+on the outside modules anywhere in `lib`.
+
+**Track depenencies**
+
+Track dependencies to be limited to:
+  * `gravitational/roundrip`,  `gravitational/trace`
+  * grpc google package and its dependencies
+  * protobuf generated code
+  * Keep some auth service specific code using `utils.Addr` and other specific code
+  *out of the `api/client` and in 
+
+
+**Go native user experience**
+
+Achieve the user experience in Client:
+
+```go
+import (
+   "github.com/gravitational/teleport/api/client"
+)
+
+func main() {
+   // TLS client is the only TLS client supported in teleport
+   client, err := client.New(client.Config{
+     Addrs: []string{"localhost:3025"},
+     // ContextDialer is optional context dialer
+     ContextDialer: net.DialContext,
+     // direct TLS credentials config
+     Credentials: client.TLSCreds(tls.Config)
+   })
+   ...
+   defer client.Close()
+
+   ping, err := client.Ping()
+   ...
+   fmt.Printf("Ping: %v\n", ping.Version)
+}
+```
+
+```
+$ go mod init
+$ go run main.go
+```
+
+**Credential providers**
+
+Support multiple credential providers with chaining
+
+* client.TLSCreds(*tls.Config) loads creds from TLS config
+* client.PathCreds("path") loads mounted creds from path, detects reloads
+and updates the grpc transport
+
+Support convenince method that supports and understands `~/.tsh/profile`,
+so users can try testing using credentials they got from `tsh login`
+
+```go
+// for testing, use client spec that loads dialer, and credentials from profile
+cfg, err := client.LoadTSHConfig()
+// cfg.ContextDialer sets up dialer using tunnel dialer in tctl
+// TLS creds are loaded from key store, similar how tctl does it in 5.0
+
+client, err := client.New(cfg)
+```
+
+```bash
+$ tsh login
+# try client
+$ go run main.go
+```
+
+**Auto version checks**
+
+Client should verify that it's implementation is compatible with
+teleport's server version by calling `Ping` anc checking the version
+similar to tsh logic.
+
+## Move Access requests code into client
+
+Move `access` workflows code into `client` subpackage:
+
+https://github.com/gravitational/teleport-plugins/tree/master/access
+
+Instead of a separate client becomes a subpackage using the only client:
+
+```go
+import (
+   "github.com/gravitational/teleport/api/client"
+   "github.com/gravitational/teleport/api/client/access"
+)
+
+func main() {
+   client, err := client.New(...)
+   watcher, err := access.NewWatcher(ctx, client, access.Watch{State: ...})
+   ...
+   defer watcher.Close()
+
+```
+
+## Provide simplified high level access hooks
+
+Access hooks are too low level, add simple handler similar to http server design
+with a router:
+
+```go
+
+func myTeamRequest(ctx context.Context, req access.Request) (access.Request, error) {
```

Comment:
can you clarify this handler function - how should it be used in the code below?

Reply:
Sure, good point. If the handler returns error, the request will be processed and that's a handler error. If the request returns modified access request it will be updated.

So approving the request would look like:

```go
func myTeamRequest(ctx context.Context, req) (...) {
   req.State = requests.StateApproved
   return req, nil
}
```